src/data/finsum.lean

```diff
@@ -0,0 +1,657 @@
+lemma finsum_in_insert (f : α → M) (h : a ∉ s) (hs : s.finite) :
+  ∑ᶠ i in insert a s, f i = f a + ∑ᶠ i in s, f i :=
+begin
+  rw [finsum_in_eq_finset_sum''' f (set.finite.insert a hs), finsum_in_eq_finset_sum''' f hs,
+      set.finite.to_finset, set.finite.to_finset, ← finset.sum_insert],
+    { congr, ext, simp },
+    { rw set.mem_to_finset, exact h }
+end
+
+/-- A more general version of `finsum_in_insert` that requires `s ∩ function.support f` instead of
+  `s` to be finite. -/
+@[simp] lemma finsum_in_insert' (f : α → M) (h : a ∉ s) (hs : (s ∩ function.support f).finite) :
+  ∑ᶠ i in insert a s, f i = f a + ∑ᶠ i in s, f i :=
+begin
+  by_cases ha : a ∈ function.support f,
+    { have := finsum_in_insert f
+        (show a ∉ s ∩ function.support f, by exact λ h', h h'.1) hs,
+      rw [← finsum_in_inter_support] at this,
+      rw [← this, finsum_in_inter_support],
+      congr, rw [set.insert_inter, set.insert_eq_of_mem ha] },
+    { rw [finsum_in_inter_support,
+          set.insert_inter_of_not_mem ha,
+         ← finsum_in_inter_support, function.nmem_support.1 ha, zero_add] }
+end
+
+/-- If `f a = 0` for all `a ∉ s`, then the sum on `insert a s` over the function `f` equals the sum
+  on `s` over `f`. -/
+lemma finsum_in_insert_of_eq_zero_if_not_mem (h : a ∉ s → f a = 0) :
+  ∑ᶠ i in (insert a s), f i = ∑ᶠ i in s, f i :=
+begin
+  by_cases hs : ((s ∩ function.support f).finite),
+    { by_cases hm : a ∈ s,
+        { simp_rw set.insert_eq_of_mem hm },
+        { rw [finsum_in_insert' f hm hs, h hm, zero_add] } },
+    { rw [finsum_in_eq_zero_of_infinite hs, finsum_in_eq_zero_of_infinite],
+      intro hs', apply hs,
+      by_cases hm : a ∈ function.support f,
+        { apply set.finite.subset hs',
+          finish },
+        { rwa set.insert_inter_of_not_mem hm at hs' } }
+end
+
+/-- If `f a = 0`, then the sum on `insert a s` over the function `f` equals the sum on `s` over `f`.
+  -/
+@[simp] lemma finsum_insert_zero (h : f a = 0) :
+  ∑ᶠ i in (insert a s), f i = ∑ᶠ i in s, f i :=
+finsum_in_insert_of_eq_zero_if_not_mem (λ _, h)
+
+/-- The sum on a singleton `{a}` over the function `f` is `f a`. -/
+@[simp] lemma finsum_singleton : ∑ᶠ i in {a}, f i = f a :=
+begin
+  rw [finsum_in_eq_finset_sum''' f (set.finite_singleton a), set.finite.to_finset,
+      ← @finset.sum_singleton _ _ a f _],
+  congr, ext, simp
+end
+
+/-- The sum on the set of two unique elements `{a, b}` over the function `f` equals `f a + f b`. -/
+@[simp] lemma finsum_pair (h : a ≠ b) : finsum_in {a, b} f = f a + f b :=
+begin
+  rw [finsum_in_eq_finset_sum''' f (show set.finite {a, b}, by simp),
+      set.finite.to_finset, ← finset.sum_pair h],
+  congr, ext, simp
+end
+
+/-- The sum on the image `g '' s` over the function `f` equals the sum on `s` over `f ∘ g` given `g`
+  is injective on `s`. -/
+@[simp] lemma finsum_in_image' {s : set β} {g : β → α} (hs : s.finite) (hg : set.inj_on g s) :
+  ∑ᶠ j in (g '' s), f j = ∑ᶠ i in s, (f ∘ g) i :=
```

Comment:
```suggestion
  ∑ᶠ j in (g '' s), f j = ∑ᶠ i in s, f (g i) :=
```

Reply:
Can this be inlined into the next proof, if it's a strict generalization? Either way it probably shouldn't be a simp lemma if the one below always fires.

Reply:
Done